src/ui/public/query_bar/directive/query_bar.js

```diff
@@ -15,55 +20,106 @@ module.directive('queryBar', function () {
       query: '=',
       appName: '=?',
       onSubmit: '&',
-      disableAutoFocus: '='
+      disableAutoFocus: '=',
+      indexPatterns: '='
     },
     controllerAs: 'queryBar',
     bindToController: true,
-    controller: callAfterBindingsWorkaround(function ($scope, config, PersistedLog) {
+
+    controller: callAfterBindingsWorkaround(function ($scope, $element, $http, $timeout, config, PersistedLog, indexPatterns) {
       this.appName = this.appName || 'global';
       this.availableQueryLanguages = queryLanguages;
       this.showLanguageSwitcher = config.get('search:queryLanguage:switcher:enable');
 
+      let persistedLog;
+
+      this.getIndexPatterns = () => {
+        if (compact(this.indexPatterns).length) return Promise.resolve(this.indexPatterns);
+        return Promise.all([indexPatterns.getDefault()]);
+      };
+
       this.submit = () => {
         if (this.localQuery.query) {
-          this.persistedLog.add(this.localQuery.query);
+          persistedLog.add(this.localQuery.query);
         }
         this.onSubmit({ $query: this.localQuery });
+        this.suggestions = [];
       };
 
       this.selectLanguage = () => {
         this.localQuery.query = '';
         this.submit();
       };
 
-      this.onTypeaheadSelect = (item) => {
-        this.localQuery.query = item;
-        this.submit();
+      this.suggestionTemplate = suggestionTemplate;
+
+      this.handleKeyDown = (event) => {
+        if (['ArrowLeft', 'ArrowRight', 'Home', 'End'].includes(event.key)) {
+          this.updateSuggestions();
+        }
       };
 
-      this.updateTypeaheadItems = () => {
-        const { persistedLog, localQuery: { query } } = this;
-        this.typeaheadItems = persistedLog.get().filter(recentSearch => {
-          return recentSearch.includes(query) && recentSearch !== query;
-        });
+      this.updateSuggestions = () => {
+        const { query } = this.localQuery;
```

Comment:
It's possible for `query` to be undefined here, and get passed into the suggestion code, which will throw. This needs a check. You can recreate the issue by removing the `query.query` part section in the URL, which modifies appState, and thus `this.localQuery`.

Reply:
It'll also throw another error if you set `query:()` in the appState. The resulting `_msearch` fails. Here's an example query in case it's helpful.

```
{"index":["logstash-*"],"ignore_unavailable":true,"preference":1521590260741}
{"version":true,"size":500,"sort":[{"@timestamp":{"order":"desc","unmapped_type":"boolean"}}],"_source":{"excludes":[]},"aggs":{"2":{"date_histogram":{"field":"@timestamp","interval":"30m","time_zone":"America/Phoenix","min_doc_count":1}}},"stored_fields":["*"],"script_fields":{},"docvalue_fields":["@timestamp","relatedContent.article:modified_time","relatedContent.article:published_time","utc_time"],"query":{"bool":{"must":[{},{"range":{"@timestamp":{"gte":1520751600000,"lte":1520818649832,"format":"epoch_millis"}}}],"filter":[],"should":[],"must_not":[]}},"highlight":{"pre_tags":["@kibana-highlighted-field@"],"post_tags":["@/kibana-highlighted-field@"],"fields":{"*":{}},"fragment_size":2147483647}}
```